pkg/platform/packet/packet.go

```diff
@@ -288,6 +294,46 @@ func createTerraformConfigFile(cfg *config, terraformPath string) error {
 	return nil
 }
 
+func validateNodePrivateCIDRs(cfg *config) error {
+	// If both the deprecated and the new fields are provided then error out.
+	if cfg.NodePrivateCIDR != "" && len(cfg.NodePrivateCIDRs) > 0 {
+		//nolint:stylecheck
+		return fmt.Errorf("`node_private_cidr` and `node_private_cidrs` cannot be used together." +
+			" Use `node_private_cidrs` only.")
+	}
+
+	// If just the deprecated field is provided then produce a warning.
+	if cfg.NodePrivateCIDR != "" {
+		fmt.Printf("Warning: `node_private_cidr` is deprecated, use `node_private_cidrs` instead.\n")
+
+		cfg.NodePrivateCIDRs = []string{cfg.NodePrivateCIDR}
+
+		// createTerraformConfigFile is called multiple times (once during the `terraform init` and
+		// during `terraform apply`), in turn this function is called multiple times. If we make
+		// changes to the object `cfg` once it is fine but when called again and if we make
+		// modifications then these checks give wrong results. If a user sets only
+		// `node_private_cidr` and here we save it to `NodePrivateCIDRs` and if this function is
+		// called twice then we have data in two places in `NodePrivateCIDR` and in
+		// `NodePrivateCIDRs` so this function fails the second time at the first condition of this
+		// function.
+		cfg.NodePrivateCIDR = ""
```

Comment:
Why can't we do this selection in the template itself? Then we don't have to do any mutation and add hacks like this.

Reply:
Because `node_private_cidr` is a deprecated field which is due to removal in another two releases. And adding complex logic in template only makes it unreadable.

Reply:
The deeper question is why these vars are initialised twice? As shown here: https://github.com/kinvolk/lokomotive/pull/1359#discussion_r579279619.

Reply:
>Because node_private_cidr is a deprecated field which is due to removal in another two releases. And adding complex logic in template only makes it unreadable.

I'd rather have a bit of if else in template rather than a function which mutates my struct and have 10 lines of comments...

>The deeper question is why these vars are initialised twice? As shown here: #1359 (comment).

If we are idempotent, this can be called as many times in whatever order we want...

Reply:
To maintain idempotency we can pass new variable like other lists do. https://github.com/kinvolk/lokomotive/blob/1d7f0e073e935ac2ea25d9fa5ab6c1dce3a1e67e/pkg/platform/packet/packet.go#L271-L281

Reply:
I think converting slices to strings before entering the template shows lack of knowledge about the templating engine. Structs define how data is organized, templates controls how they will look like when printed.

We could also have a dedicated function for rendering string slices into Terraform if that's a common feature.

Reply:
I did not mean we send strings, I said we can pass "a new variable".

Reply:
Ah, I get what you mean now. I'm not a fan of this anonymous template structs, but I guess that would work :+1:

Reply:
Done.